MuvrKit/MKWatchConnectivity.swift

```diff
@@ -23,6 +23,81 @@ struct MKConnectivitySettings {
     }
 }
 
+struct MKConnectivitySessions {
+    private var sessions: [MKExerciseSession: MKExerciseSessionProperties] = [:]
+    private static let lock: NSObject = NSObject()
+    
+    mutating func update(session: MKExerciseSession, propsUpdate: MKExerciseSessionProperties -> MKExerciseSessionProperties) {
+        objc_sync_enter(MKConnectivitySessions.lock)
+        if let oldProps = sessions[session] {
+            let newProps = propsUpdate(oldProps)
+            if oldProps.ended && !newProps.ended { fatalError("Session resurrection") }
+            sessions[session] = propsUpdate(oldProps)
+        }
+        objc_sync_exit(MKConnectivitySessions.lock)
+    }
+    
+    mutating func update(session: MKExerciseSession, newProps: MKExerciseSessionProperties) {
```

Comment:
Here we expose the same issues as we had before as ``newProps`` may contain outdated values .... except that we now check for the invariants